### DISCUSSION TITLE
HOWTO maintaining build artifacts for subsequential workflow runs ?❓ 🧐

### DISCUSSION THREAD

**Discussion by ipatch** [Homebrew/discussions#2238] [Tap maintenance and Homebrew development]
i've been working with the [homebrew-freecad][lnk1] tap off and on for several weeks now. recently i'm working on adding new versioned formula to the tap, specifically [opencascade@7.5.3][lnk2] i'm using self-hosted runners for mojave, catalina, and bigsur to build bottles for the formula contained within the tap.

however i've recently stumbled upon the [_re run all jobs_][lnk3] annoyance, [and][lnk4], 🤦‍♀️ when using github actions. this known issue becomes a super PITA if a job on one particular OS passed but on another OS it fails, thus causing all subsquent jobs to rerun. (one particular use case i came across was having an existing `_work` directory thus causing the job to fail until i manually removed unnecessary files, then the job passed, but the already completed job had to be rerun as well). when building software that can take ~ an hour to build one formula for one particular OS things can become cumbersome quickly.

scrolling through above mentioned github issue, and github community discussion, i saw one particular "work around", and i attempted to implement it with the `tests.yml` file for the homebrew-freecad tap, but unfortunately [my current setup][lnk5] is still building the entire software from scratch and doesn't seem to pick up on previous build artifacts.

i'm open to suggestions on how i can maintain successful builds without having to rebuild for all OSes contained within the matrix, and i'm aware i could split the `tests.yml` file into OS specific workflow files, but that seems to introduce a lot of redundant code, and really like to keep things simple if possible. i also could see this being a possible PR going upstream for people creating new taps, and hopefully speeding rebuilds.

[lnk1]: <github.com/freecad/homebrew-freecad>
[lnk2]: <https://github.com/FreeCAD/homebrew-freecad/pull/242>
[lnk3]: <https://github.community/t/ability-to-rerun-just-a-single-job-in-a-workflow/17234>
[lnk4]: <https://github.com/actions/runner/issues/432#issuecomment-834271999>
[lnk5]: <https://github.com/FreeCAD/homebrew-freecad/blob/master/.github/workflows/tests.yml>

**Comment by Bo98** [comment]
We don't have a solution for this ourselves - we have the same experience in Homebrew-core.

In terms of your cache, it seems like it only restores if the date & time matches, which should never be the case. You probably want a partial restore key like `${{ github.run_id }}-${{ matrix.os }}-` which should match anything with that prefix.

Instead of a date, you can also do `${{ github.run_attempt }}` nowadays (though again you wouldn't want this in the restore key since you want to restore from a different attempt).

**Reply by ipatch** [reply]
yeah i initially had `${{ github.run_id }}-${{ matrix.os }}-` but appended the date thinking it was required. i will try without and see if it helps. thanks for pointing that out.
